File: rd-framework/rd-framework-jvm/src/test/kotlin/com/jetbrains/rider/framework/test/cases/RdAsyncTaskTest.kt
package com.jetbrains.rider.framework.test.cases

import com.jetbrains.rider.framework.*
import com.jetbrains.rider.framework.base.static
import com.jetbrains.rider.framework.impl.RdCall
import com.jetbrains.rider.framework.impl.RdEndpoint
import com.jetbrains.rider.framework.impl.RdOptionalProperty
import com.jetbrains.rider.framework.impl.RdTask
import com.jetbrains.rider.util.reactive.hasValue
import com.jetbrains.rider.util.reactive.valueOrThrow
import com.jetbrains.rider.util.threading.Linearization
import kotlin.concurrent.thread
import kotlin.test.*

class RdAsyncTaskTest : RdTestBase() {
    @Suppress("UNCHECKED_CAST")
    @Test
    fun TestDynamic() {
        val property_id = 1
        val client_property = RdOptionalProperty(RdCall as ISerializer<RdCall<Int, String>>).static(property_id)
        val server_property = RdOptionalProperty(RdEndpoint.Companion as ISerializer<RdEndpoint<Int, String>>).static(property_id).slave()


        //bound
        clientProtocol.bindStatic(client_property, "top")
        serverProtocol.bindStatic(server_property, "top")

        server_property.set ( RdEndpoint(Int::toString) )


        assertEquals("1", client_property.valueOrThrow.sync(1))

        val l = Linearization()
        server_property.set (
            RdEndpoint<Int, String> { _, v ->
                RdTask<String>().apply {
                    thread {
                        l.point(1)
                        this.set(v.toString())
                        l.point(2)
                    }
                }
            }
        )

        l.point(0)
        assertEquals("2", client_property.valueOrThrow.sync(2))
        l.point(3)
        l.reset()


        //wait for task
        val task = client_property.valueOrThrow.start(3)
        assertFalse(task.isSucceeded)
        assertFalse(task.isCanceled)
        assertFalse(task.isFaulted)
        assertFalse(task.result.hasValue)
        l.point(0)

        l.point(3)
        assertTrue(task.isSucceeded, "${task.result.hasValue}")
        assertFalse(task.isCanceled)
        assertFalse(task.isFaulted)
        assertTrue(task.result.hasValue)

        l.reset()
        val interruptedTask = client_property.valueOrThrow.start(0)
        //terminate lifetime by setting another endpoint
        server_property.set( RdEndpoint(Int::toString))
        assertTrue { interruptedTask.isCanceled }
        l.disable()

        when (interruptedTask.result.valueOrThrow) {
            is RdTaskResult.Success<*> -> assertFails {  }
            is RdTaskResult.Cancelled<*> -> {}
            is RdTaskResult.Fault<*> -> assertFails {  }
            else -> assertFails {  }
        }
    }

}

File: rd-cpp/src/rd_framework_cpp/src/test/cases/RdAsyncTaskTest.cpp
#include "gtest/gtest.h"

#include "RdFrameworkTestBase.h"

#include "RdProperty.h"
#include "RdCall.h"
#include "RdEndpoint.h"

#include "Linearization.h"

using namespace rd;
using namespace rd::util;
using namespace test;

TEST_F(RdFrameworkTestBase, DISABLED_TestDynamic) {
	RdProperty<RdCall<int32_t, std::wstring> > client_property;
	RdProperty<RdEndpoint<int32_t, std::wstring> > server_property;

	statics(client_property, static_entity_id);
	statics(server_property, static_entity_id);
	server_property.slave();

	bindStatic(clientProtocol.get(), client_property, static_name);
	bindStatic(serverProtocol.get(), server_property, static_name);

	server_property.emplace([](int32_t x) -> std::wstring {
		return std::to_wstring(x);
	});

	EXPECT_EQ(L"1", client_property.get().sync(1).value_or_throw().unwrap());

	Linearization l;

	server_property.emplace([&l](Lifetime lifetime, int32_t const &v) -> RdTask<std::wstring> {
		RdTask<std::wstring> task;

		std::thread t([&l, task, v] {
			l.point(1);
			task.set(std::to_wstring(v));
			l.point(2);
		});
		t.detach();

		return task;
	});

	l.point(0);
	EXPECT_EQ(L"2", client_property.get().sync(2).value_or_throw().unwrap());
	l.point(3);
	l.reset();

	//wait for task
	auto task = client_property.get().start(3);
	EXPECT_FALSE(task.is_succeeded());
	EXPECT_FALSE(task.is_canceled());
	EXPECT_FALSE(task.is_faulted());
	EXPECT_FALSE(task.has_value());

	l.point(0);

	l.point(3);
	EXPECT_TRUE(task.is_succeeded());
	EXPECT_FALSE(task.is_canceled());
	EXPECT_FALSE(task.is_faulted());
	EXPECT_TRUE(task.has_value());

	l.reset();

	auto interrupted_task = client_property.get().start(0);

	//terminate request
	client_property.emplace();

	EXPECT_TRUE(interrupted_task.is_canceled());
	l.disable();

	AfterTest();
}